File: src/domain/usecases/StartLessonUseCase.ts
import { ILessonRepository } from '../repositories/ILessonRepository';
import { Lesson } from '../entities/Lesson';
import { EntityNotFoundError } from '@/core/errors/DomainError';

export class StartLessonUseCase {
  constructor(private lessonRepo: ILessonRepository) {}

  async execute(lessonId: string): Promise<Lesson> {
    const lesson = await this.lessonRepo.getLessonById(lessonId);
    if (!lesson) {
      throw new EntityNotFoundError('Dars', lessonId);
    }
    return lesson;
  }
}

File: mobile/lib/domain/usecases/start_lesson_use_case.dart
import '../repositories/i_lesson_repository.dart';
import '../entities/lesson.dart';
import '../../core/errors/domain_error.dart';

class StartLessonUseCase {
  final ILessonRepository lessonRepository;

  const StartLessonUseCase(this.lessonRepository);

  Future<Lesson> execute(String lessonId) async {
    final lesson = await lessonRepository.getLessonById(lessonId);
    if (lesson == null) {
      throw EntityNotFoundError('Dars', lessonId);
    }
    return lesson;
  }
}
